Merge WIP progress with main Hybberish repo for review.

ELthomasoPostfix opened the pull request:
The sections that are currently most WIP follow, refer to the recent commit history for a bit more information:
* TM arithmetic: [header](https://github.com/ELthomasoPostfix/hybberish/blob/tm-flowpipe-overapprox/src/taylormodel/taylormodel.h) | [tests](https://github.com/ELthomasoPostfix/hybberish/blob/tm-flowpipe-overapprox/tests/taylormodel_test.c)
* TM flowpipe computation: Picard operator and TM extension of Picard operator [header](https://github.com/ELthomasoPostfix/hybberish/blob/tm-flowpipe-overapprox/src/taylormodel/tmflowpipe.h) | [tests](https://github.com/ELthomasoPostfix/hybberish/blob/tm-flowpipe-overapprox/tests/tmflowpipe_test.c)

First of all, I implemented all structs that represent vectors or lists of something as linked lists, similar to the ODEList struct. This was for simplicity and consistency, and makes multi-pointer approaches a bit cleaner (no need to mix counter variables AND next pointers).

Since TMs are intended to be exclusively derived from ODEs, the TM struct keeps a 'fun' member as well. This member is used to do sanity checks within key functions: the ordering the individual TMs in a linked list of TMs should always be preserved. A bug w.r.t. wrongly ordered linked lists would be quite insidious, so I decided this approach would be safest.

Note: The [ODE parser](https://github.com/ELthomasoPostfix/hybberish/blob/tm-flowpipe-overapprox/src/parsers/odes.l) was changed to only allow integer exponents, by splitting the NUMBER token up into FLOAT and INTEGER tokens. An additional lex+parse set of files was added for parsing variable domains, in anticipation of requiring domains as input as well.

---

An example of some commonly used structs:
ODE:  $x' = y + 1; y' = x + 1;$
TM:  $(p(x) = y + 1, [0, 0]); (p(y) = x + 1; [0, 0]);$
Domains: $x \text{ in } [-1, 1]; y \text{ in } [10, 12];$
Valuation: $x = 100; y = 200;$

The TM notation "p(x) = ..." denotes that this TM was created w.r.t. the x' component of the ODE. Keeping track of that fact is essential. Domains represents what domains the variables in expressions are restricted to. Valuation is an exact valuation of the variables (using real values) which can be used for testing as discussed last meeting.

ELthomasoPostfix committed:
Add basic float-based Interval arithmetic.

This commit contains only source code,
 not the meson build config and tests.
 This is done intentionally to make the
 commits more modular and self-contained.

The current implementation is fully float
 based and provides limited interval
 arithmetic as defined in the Flow* thesis.

The interface is the following:
  * A constructor 'newInterval' to enforce
    well formedness
  * binary operations: +,-,*,/,=,in where
    'in' refers to 'I1 subset of I2' for
    intervals, not checking if a float is
    in an interval
  * interval properties: Width(I), Mid(I),
    Mag(I) and a degeneracy test.
==> Note that some operations require an
  epsilon to allow flexibility in approximation
  errors for equality comparisons.

ELthomasoPostfix committed:
Add Interval meson config.

This config will later be updated to
  accommodate for new source files.

Note: The gcc '-lm' flag is explicitly
  specified in the config due to linking
  errors on my machine.
==> See if this can be resolved another way?

ELthomasoPostfix committed:
Add Interval tests.

Add tests for:
  * binary interval arithmetic
  * interval properties
  * interval printing function

Additional tests for edge cases
  could be necessary.

The meson config explicitly specifies
  the gcc '-lm' flag due to linking
  errors on my machine.

ELthomasoPostfix committed:
Add missing Interval #ifndef.

ELthomasoPostfix committed:
Add WIP expression tree transformations.

Transformations are functions that
 generate a new, changed version of
 a given expression tree that has
 desirable properties.
 e.g. resolving neutral and absorbing
   elements of operators +,*, ...

Not all functions declared in the header
 file are implemented. Currently only
 simplification is partially implemented:
  * resolve all neutral elements for all operators
  * resolve all absorbing elements for all operators
The current implementation always works fully bottom up,
 which leads to simpler code at the cost of
 performance (mem allocations, intermediate mem frees, ...).

This push contains a dual implementation of this
 partial simplification, for the sake of versioning:
  * OLD: simplify = simplifyNeutralTerms + simplifyAbsorbingTerms
  * NEW: simplify = simplifyOperators

ELthomasoPostfix committed:
Remove old/split simplification function.

Replace the use of the old/split
 version of the simplify function,
 in favor of the combined version.

The problem with the old version, is
 that using the old version came down
 to calling a chain of 'simplify'
 calls which was unwieldy and never
 guaranteed that the result was
 irreducible to any other more
 simplified expression.

The new version reduces the given
 expression to an irreducible expression
 after a single call of 'simplify'.

ELthomasoPostfix committed:
Update fun/ meson build config for transformations.

ELthomasoPostfix committed:
Add transformation (exp simplify) tests.

ELthomasoPostfix committed:
Add WIP list of taylor polynomials.

The TPList struct is essentially a carbon
 copy of the ODEList struct, since both
 are lists of expressions w.r.t. some var.

This struct will either be part of a Taylor
 Model struct or be reworked/renamed to become
 the TM struct.

The inclusion of a 'char *fun' member
 allows to explicitly state that the
 contained expression is related to
 a specific variable.
 i.e. in the context of Taylor polynomial
 generation, when we compute a Lie derivative
 of the 'ExpTree *exp' member, this is the
 derivative w.r.t. a specific variable in the
 ODE system of the related vector field.

ELthomasoPostfix committed:
Add WIP Taylor mode flowpipe computation.

Implements Lie derivatives to compute the
 Taylor polynomials w.r.t. a given ODE system.

Currently does NOT use truncation yet.

NOTE: The current implementation of 'derivative'
 in 'funexp.c' leads to a memory error in the
 'lieDerivativeTPList' function of 'tmflowpipe.c'.
 This is because the derivative function does not
 make copies in the EXP_EXP_OP case of the base
 and exponent branches, while the lie derivative
 function assumes that it does. This leads to
 freeing memory twice, since multiple trees share
 the same branch objects.

ELthomasoPostfix committed:
Update Taylor Polynomial meson config.

ELthomasoPostfix committed:
Add exact tree equality check.

By exact, it is meant that
 the tree structure and contents
 must match exactly.
 i.e. tree  ((a + b) + c)
 is equal to itself, but not to
 (a + (b + c)) nor to
 ((d + b) + c)

ELthomasoPostfix committed:
Add very basic ExpTree equality tests.

Several concepts are tested naively:
  * Basic leaf nodes are equal to themselves
  * More complex expressions are equal to themselves
  * Differring basic leaf nodes are not equal
  * More complex, semantically same but structurally
    differing expressions are not equal

ELthomasoPostfix committed:
Fix: copy subtrees in derivative EXP_EXP_OP case.

The EXP_EXP_OP case of the derivative function
 previously did not copy the base and exponent
 subtrees before using them to construct the
 derivative. This lead to freeing memory twice.
==> Now copy base and exponent before use

ELthomasoPostfix committed:
Add TM flowpipe pre-condition.

The Taylor poly order must be
 lte to the truncation order,
 since terms exceeding the
 truncation order would be
 truncated anyways, even if the
 polynomial computation process
 generated them intermediately.

ELthomasoPostfix committed:
Implement 'toSumOfProducts(ExpTree*)'.

This transformation applies the
 mult's (*) distributive property
 w.r.t. add (+) and sub (-) at
 all levels of the tree, recursively.
==> The output should not be algebraically
 reducible any more after a single call.

The result should be a sum of products,
 where the output tree's structure mirrors
 that of the input tree in how the + and -
 operators were ordered before transformation.

NOTE: The output is a copy, it does NOT
 reorder the input tree in-place.

ELthomasoPostfix committed:
Add 'toSumOfProducts(ExpTree*)' tests.

Tests were added for the following general cases:
  * single term distribution:
    a(x + y) = ax + ay
 * distributive term distribution:
    (a + b)(x - y) = ax - ay + bx - by
 * other, complex cases:
  >  a * (expression) * b
  > a case with a deeply nested,
  complex distributive term and
  another outer, relatively complex
  distributive term
  ==> tests proper distribution on
  multiple levels in a large equation
  >  x * sin(a(y + z)) = x * sin(ay + az)

Some edge cases may remain untested; this
 suite was meant to test the broad strokes.

ELthomasoPostfix committed:
Remove unused vars in transformations tests.

ELthomasoPostfix committed:
Rename params of funexp isEqual().

ELthomasoPostfix committed:
Add monomial term degree computation.

Add a basic function to compute
 the degree of a monomial expression.
 A monomial expression is a single
 term from a polynomial, so may only
 contain vars, nums, mult (*), unary
 negative (-) and exp (^),
 e.g. -((2 * (x^2)) * y)

ELthomasoPostfix committed:
meson config: explicitly link C math library.

ELthomasoPostfix committed:
Add tests for monomial degree computation.

The test file is somewhat misleadingly
 named. It is inteded to contain tests
 related to funexp trees that are
 polynomials. Since a monomials are
 atomary elements of polynomials,
 they are polynomials themselves.

ELthomasoPostfix committed:
clang-format code + change workflow bison&flex PATH addition.

ELthomasoPostfix committed:
Update meson-actions.yml

Allow manual triggering of the meson actions workflow.

ELthomasoPostfix committed:
Remove legacy function declaration.

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

Update the PATH and set flags to point the macOS default C compiler (clang) to the installed versions of flex and bison, as opposed to the default installations on macOS. Hopefully.

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Remove pesky -Wstrict-prototypes warnings on macOS.

ELthomasoPostfix committed:
Merge branch 'tm-flowpipe-overapprox' of https://github.com/ELthomasoPostfix/hybberish into tm-flowpipe-overapprox
Merge github workflows edits from upstream.

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
Update meson-actions.yml

ELthomasoPostfix committed:
clang-format LLVM style + clean workflow yml.

ELthomasoPostfix committed:
Merge branch 'tm-flowpipe-overapprox' of https://github.com/ELthomasoPostfix/hybberish into tm-flowpipe-overapprox
Merge from upstream fork/tm-flowpipe-approx.

ELthomasoPostfix committed:
Add unary neg (-) distribution to toSumOfProducts().

Add a new function to transformations.c
 that distributes a unary neg (-a) across
 a given expression tree.
 e.g. -(-a - b) = (a + b)

Update the toSumOfProducts NEG case
 to make FIRST recursively call
 toSumOfProducts of the expression
 without the NEG operator, and THEN
 distribute the NEG over the result.
 e.g. suppose exp = -(a * (x + y))
  1) res = toSumOfProducts((a * (x + y)))
  2) distribute - over res == (a * x) + (a * y)
  3) final result is  (-(a * x) - (a * y))

Note, toSumOfProducts does NOT produce
 a canonical form:
  -(a * (x + y))  => (-(a * x) - (a * y))
  (a * -(x + y)) => ((a * -x) - (a * y))

Now that I think about it, this implementation
 is pretty inefficient, since subsequent calls
 of distributeNeg are called on growing subtrees
 of the input tree. But hey, it works for now.

Added tests for the added distributeNeg() func,
 and for the integration of that func into
 toSumOfProducts. Updated existing tests to handle
 these changes as well.

ELthomasoPostfix committed:
Apply clang-format LLVM style.

ELthomasoPostfix committed:
Apply clang-format LLVM style.

ELthomasoPostfix committed:
Add function to substitute a var by an exp.

Add a function to substitute in
 some source tree all instances of
 a given variable (specified by name),
 by a given target tree.

This function does NOT work in-place,
 instead a new instance with all variable
 instances replaces is returned.

ELthomasoPostfix committed:
Add tests for new substitution function.

ELthomasoPostfix committed:
Add first truncation implementation.

This first implementation of Trunck
 from the Flow* thesis implicitly
 assumes that the input is a polynomial/
 sum of products. Using this assumption,
 it recurses from the root down any ADD/SUB/NEG
 operators and treats all other subtrees/leaves
 as monomial terms, which to prune if their
 degree exceeds the given threshold k.

If the input is not in the form of a poly/sum of
 prods, then the output will generally not be
 pruned as intended.

BUT, there is no explicit check or transformation
 inside the truncation function that guarantees
 such a property.

ELthomasoPostfix committed:
Add minimal tests for the new truncate functions.

ELthomasoPostfix committed:
Fix: copy base & exponent in integral EXP_EXP_OP case.

ELthomasoPostfix committed:
Add definite integral function.

The implementation relies on the
 substitution transformation.
 It creates a SUB node of the
 substitutions of the upper & lower
 bounds of the indefinite integral.
 i.e. int_a^b (x)
    = int(x)[x := b] - int(x)[x := a]
 where int(x) represents the indefinite
 integral and a & b are resp. the lower
 and upper bounds. The notation [x := y]
 represents substituting x for y in the
 subjected expression.

Note, a forward declaration of 'substitute'
 was added to the funexp header to avoid
 circular includes with the transfomations header.

ELthomasoPostfix committed:
Add definite integral tests.

Note, all test cases for the
 indefinite integral were copied
 from the indefinite integral case,
 and then adjusted as necessary.

BUT, the EXP_FUN cases were dropped
 entirely, due to issues porting them
 to ExpTree instances, instead of plain
 string comparison.
==> These cases should be added as soon
 as possible!

ELthomasoPostfix committed:
clang-format definite integral tests.

ELthomasoPostfix committed:
Edit comments + add extra pre-cond. asserts for TPList impl.

ELthomasoPostfix committed:
Add const qualifiers + use double, not float for Interval.

ELthomasoPostfix committed:
Finish Interval asserts + Fix incorrect DIV assert.

Add 'not NULL' assertions for all interval
 operations.
Add 'eps >= 0' assertions; negative epsilon
 values could produce unexpected results.

Fix: the DIV interval operation should not
 allow 0 to be part of the divisor. i.e.
  [a, b] / [c, d]  =>  0 not in [c, d]
 This was not correctly expressed by the
 existing assertion. Fixed now.
==> Also commented the tests that failed
 due to this changed assertion.
==> In the future, do proper error code
 checking.

ELthomasoPostfix committed:
Add Interval sqrt & unary neg + tests.

Add functions to compute the additive
 inverse (unary negative) and sqrt of
 an interval.

Add additional tests to cover the new cases.

ELthomasoPostfix committed:
clang-format LLVM style.

ELthomasoPostfix committed:
Refactor TPList into TaylorModel.

The old TPList represented a list
 of functions, that were taylor polynomials
 derived from a given system of ODEs via the
 methods described in the Flow* thesis:
 Lie derivatives or Picard iteration.

This struct has been extended to represent
 a full Taylor model: a Taylor polynomial +
 a remainder interval. The workflow is intended
 to be as follows:
   * 'computeTaylorPolynomial' produces a Taylor
   Model with a dummy remainder interval
   * Another function that implements step 2
     generates the matching safe remainder interval
     on top of those Taylor polynomials.

Note, the current implementation for TaylorModel
 may be incompatible with the full scope of the
 algorithm in Flow* (algo 6); those considerations
 are still to come.

ELthomasoPostfix committed:
Rework truncation: support returning truncated terms.

The truncation transformation function has
 been reworked to support returning a copy
 of all terms that were truncated, as an
 expression tree that is a sum of those terms.
 i.e. given k=2, we truncate
     x + x^2 - x^3 + x^4
 into
     x + x^2
 with truncated terms
     -x^3 + x^4

The code of the original 'truncate' function
 was generalized to support the extended behavior,
 as the 'truncateTerms' function.
The new interface for truncation are the 'truncate'
 and 'truncate2' functions:
  * 'truncate' has the same outward behavior as before
   * 'truncate2' also returns a copy of the truncated
     terms, as (more or less) a sum of those terms,
     through its third argument (a ExpTree** type)

ELthomasoPostfix committed:
Apply clang-format LLVM style.

ELthomasoPostfix committed:
Use the Interval 'constructor' within interval arithmetic.

Due to an oversight, the Interval
 'constructor' function, which enforces
 the well-formedness of the created
 interval, was not yet used within the
 interval arithmetic function of interval.c
Now it is used.

ELthomasoPostfix committed:
Add Interval EXP (^) operation.

The interval exponentiation
 operation for natural number
 exponents is added.

Two versions are added:
  * 'powInterval' which implements
    'Algorithm 1' from the Flow* thesis.
  * 'pow2Interval' which is implemented
    by reducing the exponentiation to a
   chain of interval multiplications
The first implementation  produces more strict
  bounds in some cases (e.g. if the interval
  [a, b] has a < 0 <= b).
The second implementation was added for
 comparative analysis (read, just in case).

ELthomasoPostfix committed:
Split off math code into a separate 'varmath' lib.

To improve modularity, split off
 files into the 'varmath' directory
 (and lib/inc in the meson config)
  * 'interval.h/c' : custom interval struct
    and interval arithmetic operations
  * 'variables.h/c' : struct(s) to do with
    valuations (interval- or float valued)
    of variables, for use in expression tree
    evaluation.

Adjust the meson config as follows:
  * The top-level meson.build now
    specifies a 'varmath_inc' subdir
  * The 'varmath' subdir meson.build
    now specifies a 'varmath_lib' lib
  * All depending meson.build files now
  include this inc and lib as dependencies

ELthomasoPostfix committed:
clang-format LLVM style previous commit.

ELthomasoPostfix committed:
Edit interval math.

Rename 'inInterval' to 'subeqInterval',
 to better reflect the subset or equal
 operation.

Add the 'elemInterval' operation. This
 implemented the element of operation,
 i.e. 0 element of [-1, 1] is true.

Modify & add the relevant tests.

ELthomasoPostfix committed:
Remove duplicate strdup.

The 'newExpLeaf' implementation
 already calls strdup on the
 char* parameter, so do not call
 it before that to prevent mem leaks.

ELthomasoPostfix committed:
Enforce integer exponents: a^n with 'n' an integer.

Split the NUMBER token into
 a FLOAT and an INTEGER token.

Add a new parser rule 'number'
 that is either a FLOAT or a
 INTEGER.

ELthomasoPostfix committed:
Add lexer+parser for var domains.

A var domain takes the following form:
  x in [-1, 1]; y in [1, 1];

Add a lexer (.l) and parser (.y) for
 such a grammar, based on the existing
 ODE lexer+parser code.

Edit the meson.build to include this
 addition.

ELthomasoPostfix committed:
Add untested var valuation struct.

A variable valuation takes the following form:
  x = 3.1516; y = 1; z = 0.1;
though no parser for it exists.

This struct is intended to allow real-valued
 evaluation of an ExpTree for use in testing
 and plotting.

ELthomasoPostfix committed:
Add WIP TM arithmetic.

Add the following evaluation methods,
 that evaluate an expression tree based
 on type-varying arithmetic:
  * 'evaluateExpTree' : interval arithmetic
  * 'evaluateExpTreeTM' : TM arithmetic (WIP)
  * 'evaluateExpTreeReal' : real-valued arithmetic (double C type)

Implement (WIP) TM arithmetic. Not all
 operations are fully complete/correct yet.
 This part of the code base should be
 thorougly tested and documented, as
 well as rigorously/mathematically verified
 at some point in time.
  * addition (+)
  * subtraction (-)
  * multiplication (*)
  * division (/)  =>  WIP
  * additive inverse (-)
  * exponentiation (^)  ==>  WIP
  * integration/anti-derivative  ==> WIP

Add a truncation function, to separately
 handle truncating the expression of the
 TM and addingthe interval enclosure to
 the remainder:
  truncTM((p, I)) = (p - pe, Int(pe) + I)
The notation of the expression above is
 used by the flowstar paper in section 2.3.3

Do clang-format LLVM style.

ELthomasoPostfix committed:
Add var lexer+parser test.

ELthomasoPostfix committed:
Add WIP Taylor model struct tests.

These tests are intended to rigorously
 test the correctness of the TaylorModel
 struct and its interface.
==> very much WIP.

ELthomasoPostfix committed:
Add WIP Picard operator code.

Add untested 'picardOperator' that
 implements the base definition of
 a Picard operator.
 The 'substituteTaylorModel' function
 implements the substitution of
 variables in the vector fields by
 the expressions of the Taylor models.
 This is NOT the TM extension of the
 Picards operator.

Add untested 'picardOperatorTM', that
 will implement the TM extension of
 the Picard operator.

ELthomasoPostfix committed:
Add WIP tests for TM flowpipe computation.

These tests are more manually playing around with
 the functions and visually checking the results
 than any actual form of testing.

ELthomasoPostfix committed:
Test future program pipeline.

By pipeline, I mean the program
 execution from parsing to
 TM polynomial computation to
 safe remainder computation.

This test is intended to, on a
 small scale, emulate the intended
 flow of the program and catch
 inconsistencies/bugs that arise
 from future changes.
 i.e. it should catch breaking
 changes of the full program flow.

ELthomasoPostfix committed:
Add TM EXP implementation.

Unroll a TM exponentiation with
 integer powers into a sequence
 of multiplications.

ELthomasoPostfix committed:
Clean newExpLeaf + docs.

Explicitly document that newExpLeaf
 applies a strdup internally, and that
 the provided name's lifetime is not
 taken over by the leaf.

Document other ExpTree constructor
 functions similarly.

ELthomasoPostfix committed:
Require no memory leaks in MacOS job.

ELthomasoPostfix committed:
Work on TM arithmetic + mem leak fixes.

Work on the EXP (^)
 and DIV (/) TM arithmetic
 operators.

Some memory leak fixes for
 the TM truncation function.

ELthomasoPostfix committed:
Add lib for utility functions.

The 'utils' lib should contain
 extra functions that do not
 belong to any particular other
 lib, or that are used by many
 other libs.

Currently, only contains functions
 for converting other types to and
 from char* type.

ELthomasoPostfix committed:
Edit ode parser to fix mem leak.

Forcing strdup in newExpLeaf
 results in memory leaks in
 odes.y since odes.l does
 strdup on the leaf tokens.

Also edit the related tests.

ELthomasoPostfix committed:
Fix memory leaks + edit meson config.

The meson config was changed to
 incorporate the 'utils' lib where
 used. This was forgotten in a
 previous commit.

Many, many memory leaks were fixed.
 Most of them were of the following type:
  * strdup for a newExpLeaf 'name' param
  * Tests that did not properly handle
    (delete) objects
 Most occurrences happended in tests.

The other leaks were found in;
  * funexp integral
  * funexp derivative
  * taylormodel functions
  * tmflowpipe functions

Note: All struct constructors require
 their pointer arguments to be heap
 allocated or NULL, EXCEPT for newExpLeaf,
 which calls strdup internally.
 This is because every single leaf requires
 a name argument, so this makes leaf
 definitions somewhat shorter at the cost
 of more confusing usage.

ELthomasoPostfix committed:
Edit header docs for constructor functions.

ELthomasoPostfix committed:
Require no memory leaks in MacOS job.

Actually install valgrind before calling it :/

ELthomasoPostfix committed:
Require no memory leaks in MacOS job #3.

ELthomasoPostfix committed:
Require no memory leaks in MacOS job.

ELthomasoPostfix committed:
Require no memory leaks in MacOS job.

ELthomasoPostfix committed:
Require no memory leaks in MacOS job.

gaperez64 committed:
Merge branch 'main' into tm-flowpipe-overapprox

gaperez64 commented:
@ELthomasoPostfix Put the valgrind check in the linux one and not on Mac, the tests on linux seem to pass thus far.

ELthomasoPostfix committed:
Move valgrind meson test wrap from mac to linux job.

Adding a valgrind wrap to the meson test command enforces that the tests may not contain any true memory leaks at all.
Note: If memory is still accessible, but not freed, at the end of a test, then this does not count as a leak and the build will pass.

Installing valgrind on MacOS seems to be a pain, and native alternatives are likely a better option.

Solution: move the valgrind wrap on the meson test command from the linux job to the mac job.

ELthomasoPostfix committed:
clang-format LLVM style.

ELthomasoPostfix committed:
Install valgrind on linux job.

ELthomasoPostfix committed:
sudo in github actions should work for package installs?

ELthomasoPostfix commented:
> @ELthomasoPostfix Put the valgrind check in the linux one and not on Mac, the tests on linux seem to pass thus far.

I went ahead and fixed that. Didn't know that new commits automatically became part of the merge request :sweat_smile:.

Valgrind on MacOS seems to be finicky to get working, though some mix of MacOS alternatives could likely serve our purposes all the same. But I personally work on linux, hence the commit spam trying to wrangle the Mac workflow into obedience. It was more a CI/CD problem than test failing.

kasperengelen commented:
Note that I did not check whether the code is correct. I assume that any functional problems will be resolved during the meetings and case-studies.
 
All tests run. Formatting OK. All original tests still present.

Comments:

- Some tests lack comments. Every test should have some description of what that test does.
- Please do not add repeated TODO statements. One statement is sufficient and can be retrieved using tools or search.
- In general, there should be more and clearer comments.
- We should try to use Doxygen comment style. This is a nice tool and the comments are very easy to read.

Headers should have proper whitespace in between function declarations. Some stuff does not have comments.
Tests:
- Use “actual” instead of “got”. Here I refer to other languages: Python, Java, Swift where “actual” is used.
- derivative_test.c:157. Why is "polynomial" not freed here?
- odeparse_test.c: the original part has been duplicated three times. Why? Please add comments.

Typing:
- ../src/taylormodel/taylormodel.c:748:46: warning: passing 'const char *const' to parameter of type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
- Why do we have pointers that are const? I think "const X*" is useful, but "X* const" not so much. @gaperez64 any advice here?

Taylormodel header: the includes for the standard library should come before the includes of other project files.

Flowpipe header:
- /* Substitute each of the functions, w.r.t. their assigned variable, into all of the ODEs. */
- This is not clear to me.
- There are a lot of Lie-derivative functions here. The difference is not always clear. The API of the tool should be clear and easy to use.
 
The stuff above should be remediated before the merge can happen. I propose that we focus on this before working on additional features. Please ask me if anything is unclear.
 
Some other notes:
- The separate fork is a good idea. Let’s keep it that way
- Please make smaller pull requests in the future
- The EXP_FUN enum constant should be removed, and replaced with “EXP_SQRT” “EXP_SIN”, etc. Keeping stuff as enum constants is cheap and type-safe (and can be used in switch statements!)
- We should discuss during the meetings what the API should look like.

ELthomasoPostfix committed:
Add WIP TM integration python code.

The basis of this code was the Hybberish
 ipynb prof. Perez shared.

Attempts were made to generalize the entire
 tm integration code that was already present
 and then extend it to implement the entire
 tm integration routine.

This code makes use of sympy for symbolic
 computations and interval evaluation of
 those expressions. matplotlib is used
 to plot for each flowpipe an interval
 enclosure in the form of a single box.
 This implies plotting only works properly
 for 2D ODEs, which is strictly enforced
 in the current code.

ELthomasoPostfix committed:
Document all of the python TM integration function.

ELthomasoPostfix committed:
Replace function definitions by imports.

Replace all the functions defined in the
 ipynb for TM integration by imports from
 'tm_integration.py' instead. This
 keeps the notebook cleaner/shorter, while
 still allowing indivitual testing of the
 components/functions.

ELthomasoPostfix committed:
Add doxygen for interval math.

This commit functions as an initial
 attempt at writing decent doxygen
 docs, and figuring out how to add
 it to the project.

Add note on generating doxygen docs
 to the top-most README file.

The documentation written for the
 interval math was purposefully
 extensive, since it is an interface
 that is *very* unlikely to change
 at all due to its basic nature.
Additionally, the comments describe
 on a high (mathematical/theoretical)
 level what the functions are intended
 for; implementation details are better
 suited to be placed at the actual
 function definitions / source code.

The remaining doxygen docs are likely
 to be pushed in a single commit later,
 to avoid commit-spamming the log.

ELthomasoPostfix committed:
Add missing Doxyfile.

ELthomasoPostfix committed:
Add doxygen comments in all header files.

Also add some extra asserts to
 src/fun/funexp.c in delExpTree
 and printExpTree.

The math/LaTeX notation should still
 be fairly readable as plaintext.

The use of doxygen parameter references
 (\p paramName) was used somewhat
 inconsistently. Mainly because it seems
 annoying to maintain those references
 in case of refactoring.

Some function references were used though
 (@ref funName). But this seems reasonable
 since C does not support multiple functions
 with the same name (function overloading)
 and doxygen does report that a @ref could
 not be resolved.

Class/struct member documentation is usually
 kept to a minimum to avoid overcrowding the
 source code. The class/struct main documentation
 usually a bit more extensive, and specifies
 high level intuition on the class (e.g.
 what the class mathematically represents)
 and sometimes some crucial implementation
 details or invariants.

Functions usually give a description of all
 parameters, even the trivial ones, for the
 sake of consistency.
 All parameters should also specify whether
 they are [in], [out] or even [in, out]
 parameters.

There is a decent bit of duplication in the
 documentation, mainly generic @pre or @post
 conditions, return value information, or
 in the case of several functions that perform
 the same operation, but slightly differently
 (e.g. applying an operation to one ExpTree
 vs a linked list of ExpTrees)

ELthomasoPostfix committed:
Add missing doxygen docs.

* 'src/utils/utils.h' did not have a
 file-level doxygen comment.
* 'src/systems/sysode.h' was not commented
 with doxygen at all.

ELthomasoPostfix committed:
Apply neutral refactoring for tests.

No functional changes to tests:
  * Change 'got' to 'actual' in comments,
    for variable names and inside print
    statements.
  * Remove duplicate TODO statements.
    Just use a tool to find them instead :(

ELthomasoPostfix committed:
Add const qualification to many function parameters.

ELthomasoPostfix committed:
clang-format LLVM style.

ELthomasoPostfix committed:
Add & expand comments in tests.

ELthomasoPostfix committed:
doc: Add/extend some source code comments.

ELthomasoPostfix committed:
Remove python testing/working example from this branch.

See the [dedicated branch](https://github.com/ELthomasoPostfix/hybberish/tree/python-interface)
 that was just created for the
 purpose of hosting the python code.

A separate branch was made, since this
 branch's purpose is to get a working
 version of TM integration in C.
 Adding random python code on top
 of that just muddles the waters.
 Also, it the merge request is already
 large enough :/

A python interface *is* desirable though,
 if only for plotting the generated
 output.
